### PR TITLE
Substitute BootLoader.hasClassPath() to work around a JDK 11 NPE

### LINKED ISSUE #6089: BootLoader.hasClassPath() NPE during native image execution with JDK 11
**Describe the bug**

The following error happens with most (if not all) Quarkus integration tests during the native image startup (after the generation ended successfully):

```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Running Quarkus native-image plugin on GraalVM Version 19.3.0 CE
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] docker run -v /home/git/quarkus/integration-tests/kogito/target/quarkus-integration-test-kogito-999-SNAPSHOT-native-image-source-jar:/project:z --user 0:0 --rm quay.io/quarkus/ubi-quarkus-native-image:19.3.0-java11 -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -jar quarkus-integration-test-kogito-999-SNAPSHOT-runner.jar -J-Djava.util.concurrent.ForkJoinPool.common.parallelism=1 -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:+AddAllCharsets -H:EnableURLProtocols=http -H:+JNI --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-integration-test-kogito-999-SNAPSHOT-runner
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]    classlist:  31,275.92 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]        (cap):   2,406.23 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]        setup:   5,438.86 ms
21:14:11,105 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Final
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]   (typeflow):  64,753.97 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]    (objects):  39,490.06 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]   (features):   2,174.23 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]     analysis: 111,257.16 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]     (clinit):   1,960.39 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]     universe:   6,119.57 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]      (parse):  12,618.13 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]     (inline):  15,299.03 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]    (compile):  64,538.59 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]      compile: 100,775.67 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]        image:  15,565.41 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]        write:   2,620.39 ms
[quarkus-integration-test-kogito-999-SNAPSHOT-runner:27]      [total]: 273,808.74 ms
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 284609ms
[INFO]
[INFO] --- maven-failsafe-plugin:2.22.1:integration-test (default) @ quarkus-integration-test-kogito ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.quarkus.it.kogito.jbpm.OrdersProcessTestIT
Executing [/home/git/quarkus/integration-tests/kogito/target/quarkus-integration-test-kogito-999-SNAPSHOT-runner, -Dquarkus.http.port=8081, -Dtest.url=http://localhost:8081, -Dquarkus.log.file.path=target/quarkus.log]
Exception in thread "main" java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:194)
        at io.quarkus.runtime.Application.start(Application.java:87)
        at io.quarkus.runtime.Application.run(Application.java:210)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:41)
Caused by: java.lang.NullPointerException
        at jdk.internal.loader.BootLoader.hasClassPath(BootLoader.java:101)
        at java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1189)
        at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1220)
        at java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264)
        at java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299)
        at java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384)
        at org.eclipse.microprofile.config.spi.ConfigProviderResolver.loadSpi(ConfigProviderResolver.java:151)
        at org.eclipse.microprofile.config.spi.ConfigProviderResolver.loadSpi(ConfigProviderResolver.java:146)
        at org.eclipse.microprofile.config.spi.ConfigProviderResolver.loadSpi(ConfigProviderResolver.java:146)
        at org.eclipse.microprofile.config.spi.ConfigProviderResolver.instance(ConfigProviderResolver.java:119)
        at io.quarkus.runtime.generated.Config.readConfig(Config.zig:4497)
        at io.quarkus.runtime.generated.Config.createRunTimeConfig(Config.zig:2830)
        at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:52)
        ... 3 more
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.323 s <<< FAILURE! - in io.quarkus.it.kogito.jbpm.OrdersProcessTestIT
[ERROR] testRuleEvaluation  Time elapsed: 0.024 s  <<< ERROR!
org.junit.jupiter.api.extension.TestInstantiationException: TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [io.quarkus.it.kogito.jbpm.OrdersProcessTestIT]: Failed to start native image, process has exited
Caused by: java.lang.RuntimeException: Failed to start native image, process has exited
```

It seems `ClassLoaders.bootLoader()` returns `null` which shouldn't be possible from what I can see in the JDK code (but I may read it wrong). This looks like an upstream bug but I'd like some opinions about it here before creating a GraalVM issue.

**To Reproduce**
1. Build the Quarkus sources using JDK 11.
2. `./mvnw clean install -Dnative -Dquarkus.native.container-build=true -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-native-image:19.3.0-java11 --projects :quarkus-integration-test-kogito`

**Environment:**
 - Output of `uname -a`: `Linux pc-79.home 4.18.0-80.11.2.el8_0.x86_64 #1 SMP Tue Sep 24 11:32:19 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
 - Output of `java -version`: `openjdk version "11.0.5" 2019-10-15 LTS OpenJDK Runtime Environment 18.9 (build 11.0.5+10-LTS) OpenJDK 64-Bit Server VM 18.9 (build 11.0.5+10-LTS, mixed mode, sharing)`
 - GraalVM version (if different from Java): `19.3.0`
 - Quarkus version or git rev: `master`

**Additional context**
The native image build is done using a Docker image.

### PR DESCRIPTION
**Pull request by gwenneg**
Fixes #6089

### CODE REVIEW

**Comment by gwenneg** [comment]
This fixes the NullPointerException and allows the integration tests to end successfully.

@stuartwdouglas Does this look like a reasonnable fix? I'm not really in my field of expertise here so I might be doing something totally wrong.

**Comment by stuartwdouglas** [comment]
This looks reasonable but you should file a GraalVM issue so it can be fixed upstream.

**Comment by gwenneg** [comment]
Sure, I'll do it. Thanks for having a look!

**Comment by gwenneg** [comment]
GraalVM issue created: https://github.com/oracle/graal/issues/1966

**Comment by gwenneg** [comment]
@gsmet This might be the solution to all the JDK 11 native integration tests that failed to start recently. It had definitely nothing to do with the time the native image needed to start :)

**Review by gastaldi — APPROVED** [review]

File: core/runtime/src/main/java/io/quarkus/runtime/graal/BootLoaderSubstitutions.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package io.quarkus.runtime.graal;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import com.oracle.svm.core.jdk.JDK11OrLater;
+
+@TargetClass(className = "jdk.internal.loader.BootLoader", onlyWith = JDK11OrLater.class)
+final class Target_jdk_internal_loader_BootLoader {
```
Comment (gsmet) [inline core/runtime/src/main/java/io/quarkus/runtime/graal/BootLoaderSubstitutions.java]: please add a comment explaining why we added that one with a pointer to the GraalVM issue.

**Review by gsmet — CHANGES_REQUESTED** [review]

File: core/runtime/src/main/java/io/quarkus/runtime/graal/BootLoaderSubstitutions.java (hunk 1 above)
Comment (gwenneg) [reply core/runtime/src/main/java/io/quarkus/runtime/graal/BootLoaderSubstitutions.java]: I was hoping the GraalVM issue would be labeled as a bug yesterday and waited for that to include such a comment, but it's still not recognized as a bug. I'll add a comment about the issue anyway.

**Comment by jaikiran** [comment]
Hello @gwenneg, I just commented on that graal issue you opened. The graal team has fixed it in `master` branch some days back as part of https://github.com/oracle/graal/commit/68027de.

I think we will need this workaround in Quarkus until 19.3.x is released with that fix. Their commit is different from what is being proposed in this PR - they return `true` instead of `false` for `hasClassPath`. I think we should just copy over their change in that graal commit into Quarkus as a substitution to be consistent with their semantics.

**Comment by gwenneg** [comment]
I don't think we can copy the entire `Target_jdk_internal_loader_BootLoader` class as we would have to import the `ClassForNameSupport` class in Quarkus and it is an internal class which is not exported in the svm-19.3.0.jar `module-info`. But I agree we should at least return the same `true` value than the GraalVM substitution from https://github.com/oracle/graal/commit/68027de.

**Comment by gwenneg** [comment]
Actually, changing the returned value from `false` to `true` in our substitution could also cause the other `BootLoader` methods to be called (the ones that were substituted in https://github.com/oracle/graal/commit/68027de). I'll need to run an entire Quarkus native build to check if everything's OK with such a change.

**Comment by jaikiran** [comment]
> I don't think we can copy the entire Target_jdk_internal_loader_BootLoader class as we would have to import the ClassForNameSupport class in Quarkus and it is an internal class which is not exported in the svm-19.3.0.jar module-info

Hello @gwenneg, sorry, I wasn't clear. I actually meant, we should just copy over those 5 new `Substitute` methods (which don't require any internal class usage) to the `BootLoader` that were added as part of that commit in graal.

**Comment by gwenneg** [comment]
Oh, I misunderstood you first comment (which was clear enough), sorry :) That will work and I'll do it right now.

**Comment by gwenneg** [comment]
There's an intruder in the PR, I'm removing it :)

**Comment by gwenneg** [comment]
This PR needs to be tested with a full JDK 11 native build.

**PLEASE DON'T MERGE** before I say it's OK.

**Comment by gwenneg** [comment]
Thanks for your help with this @jaikiran!

File: core/runtime/src/main/java/io/quarkus/runtime/graal/BootLoaderSubstitutions.java (hunk 1 above)
Comment (gwenneg) [reply core/runtime/src/main/java/io/quarkus/runtime/graal/BootLoaderSubstitutions.java]: Done.

**Comment by gwenneg** [comment]
I've successfully tested the latest version of this PR with many JDK 11 native integration tests. I think it can be safely merged into master after being reviewed again.

**Comment by gwenneg** [comment]
I just saw a similar NPE that may be related, I'm putting the PR back to WIP to investigate what's going on.

**Comment by gwenneg** [comment]
I created #6127 to fix the similar NPE mentioned above. Both share a common root cause but require different and independent substitutions to be fixed. I think merging this PR ASAP is important because it might help us detect new issues during native tests execution, that's why I didn't treat #6127 in the same PR.

@gsmet: We're ready for a final review and merge, for real this time.

**Review by gsmet — APPROVED** [review]
Nice work, @gwenneg .

**Pull request merged**